File: LocationProvider.java
package com.eps.highusers.services;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.Priority;

import java.util.concurrent.TimeUnit;

public class LocationProvider {

    private static LocationProvider mInstance;

    private final LocationManager locationManager;
    private final FusedLocationProviderClient fusedLocationClient;
    private LocationRequest locationRequest;


    public static Long UPDATE_INTERVAL_SECS = 10L;
    public static Long FASTEST_UPDATE_INTERVAL_SECS = 5L;
    public static Long MAXIMUM_WAIT_TIME_SECS = 15L;
    public static Float SMALLEST_DISPLACEMENT_METER = 1F;

    private LocationListener networkLocationListener;
    private LocationListener gpsLocationListener;
    private LocationCallback fusedLocationCallback;

    private final LocationObserver locationObserver;

    public static LocationProvider getInstance(Context context, LocationObserver locationObserver) {
        if (mInstance == null)
            mInstance = new LocationProvider(context, locationObserver);

        return mInstance;
    }

    private LocationProvider(Context context, LocationObserver locationObserver) {
        locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
        this.locationObserver = locationObserver;
        buildLocationRequest();
        buildLocationCallBacks();
    }

    @SuppressLint("MissingPermission")
    public void startLocationUpdates(Context context) {
        if (isPermissionGranted(context)) {

            // 1- Network Provider
            boolean isNetworkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
            if (isNetworkEnabled) {
                Log.d("LocationTest" , "Network Provider is working");
                locationManager.requestLocationUpdates(
                        LocationManager.NETWORK_PROVIDER,
                        TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SECS),
                        SMALLEST_DISPLACEMENT_METER,
                        networkLocationListener);
            }

            // 2- GPS Provider
            boolean isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
            if (isGPSEnabled) {
                Log.d("LocationTest" , "GPS Provider is working");
                locationManager.requestLocationUpdates(
                        LocationManager.GPS_PROVIDER,
                        TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SECS),
                        SMALLEST_DISPLACEMENT_METER,
                        gpsLocationListener);
            }

            // 3- Fused Provider
            fusedLocationClient.requestLocationUpdates(
                    locationRequest,
                    fusedLocationCallback,
                    Looper.getMainLooper());
        } else
            Toast.makeText(context, "من فضلك قم بمنح إذن الوصول للموقع الجغرافي", Toast.LENGTH_LONG).show();
    }

    public void stopLocationUpdates() {
        locationManager.removeUpdates(networkLocationListener);
        locationManager.removeUpdates(gpsLocationListener);
        fusedLocationClient.removeLocationUpdates(fusedLocationCallback);
    }

    // Build a location request
    private void buildLocationRequest() {
        locationRequest = LocationRequest.create();
        locationRequest.setPriority(Priority.PRIORITY_HIGH_ACCURACY);
        locationRequest.setInterval(TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SECS));
        locationRequest.setFastestInterval(TimeUnit.SECONDS.toMillis(FASTEST_UPDATE_INTERVAL_SECS));
        locationRequest.setMaxWaitTime(TimeUnit.SECONDS.toMillis(MAXIMUM_WAIT_TIME_SECS));
        locationRequest.setSmallestDisplacement(SMALLEST_DISPLACEMENT_METER);
    }

    // Build the location callback object and obtain the location results //as demonstrated below:
    private void buildLocationCallBacks() {
        networkLocationListener = new LocationListener() {
            @Override
            public void onLocationChanged(@NonNull Location location) {
                    Log.d("LocationTest" , " .. Location is : " + location);
                    locationObserver.onGettingLocationUpdates(location);
            }

            @Override
            public void onStatusChanged(String provider, int status, Bundle extras) {
                LocationListener.super.onStatusChanged(provider, status, extras);
            }

            @Override
            public void onProviderEnabled(@NonNull String provider) {
                LocationListener.super.onProviderEnabled(provider);
            }

            @Override
            public void onProviderDisabled(@NonNull String provider) {
                LocationListener.super.onProviderDisabled(provider);
            }
        };

        gpsLocationListener = new LocationListener() {
            @Override
            public void onLocationChanged(@NonNull Location location) {
                Log.d("LocationTest" , " .. Location is : " + location);
                locationObserver.onGettingLocationUpdates(location);
            }

            @Override
            public void onStatusChanged(String provider, int status, Bundle extras) {
                LocationListener.super.onStatusChanged(provider, status, extras);
            }

            @Override
            public void onProviderEnabled(@NonNull String provider) {
                LocationListener.super.onProviderEnabled(provider);
            }

            @Override
            public void onProviderDisabled(@NonNull String provider) {
                LocationListener.super.onProviderDisabled(provider);
            }
        };

        fusedLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(@NonNull LocationResult locationResult) {
                Location location = locationResult.getLastLocation();
                if (location != null) {
                    Log.d("LocationTest" , " Location is : " + location);
                    locationObserver.onGettingLocationUpdates(location);
                } 
            }
        };
    }

    private Boolean isPermissionGranted(Context context) {
        return (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED ||
                ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION)
                        == PackageManager.PERMISSION_GRANTED);
    }

    public interface LocationObserver {
        void onGettingLocationUpdates(Location location);
    }
}

File: LocationProvider.kt
package com.example.newcanalcollection.managers.location


import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.os.Looper
import android.util.Log
import com.google.android.gms.location.*
import com.google.android.gms.tasks.CancellationTokenSource
import com.google.android.gms.tasks.Task
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import java.util.concurrent.TimeUnit
import javax.inject.Inject


/*
This Class is used to get Location
remember --> you need to handle 2 events (GPS is not enabled , GPS permission is not granted)
you will find that in LocationUtils Class
 */

class LocationProvider @Inject constructor(
    context: Context
) {

    private val locationManager =
        context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
    private val fusedLocationManager = LocationServices.getFusedLocationProviderClient(context)

    private var location: Location? = null

    // For Network Location Provider
    private var networkLocationListener: LocationListener? = null

    // For Network Location Provider
    private var gpsLocationListener: LocationListener? = null

    // For Fused Location API
    private var fusedLocationCallBack: LocationCallback? = null

    @SuppressLint("MissingPermission")
    fun startLocationUpdates(): Flow<Location> = callbackFlow {

        /*
        First : Initialize Callbacks
         */
        networkLocationListener = object : LocationListener {
            override fun onLocationChanged(location: Location) {
                Log.d("locationTest", "networkLocationListener : $location")
                trySend(location).isSuccess
            }

            override fun onProviderEnabled(provider: String) {}

            override fun onProviderDisabled(provider: String) {}

            // Deprecated annotation is very important to prevent crash
            // Don't delete it
            @Deprecated("Deprecated in Java")
            override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {
            }
        }


        gpsLocationListener = object : LocationListener {
            override fun onLocationChanged(location: Location) {
                Log.d("locationTest", "gpsLocationListener : $location")
                trySend(location).isSuccess
            }

            override fun onProviderEnabled(provider: String) {}

            override fun onProviderDisabled(provider: String) {}

            // Deprecated annotation is very important to prevent crash
            // Don't delete it
            @Deprecated("Deprecated in Java")
            override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {
            }
        }

        fusedLocationCallBack = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult) {
                super.onLocationResult(locationResult)
                val location = locationResult.lastLocation
                Log.d("locationTest", "fusedLocationCallBack : $location")
                if (location != null)
                    trySend(location).isSuccess
            }
        }

        /*
         Second : Get Updates from 3 sources
         */
        // 1- Network Location Provider
        val hasNetwork = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
        if (hasNetwork) {
            locationManager.requestLocationUpdates(
                LocationManager.NETWORK_PROVIDER,
                TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SEC),
                SMALLEST_DISPLACEMENT_METER,
                networkLocationListener!!
            )
        }

        // 2- GPS Location Provider
        val hasGps = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
        if (hasGps) {
            locationManager.requestLocationUpdates(
                LocationManager.GPS_PROVIDER,
                TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SEC),
                SMALLEST_DISPLACEMENT_METER,
                gpsLocationListener!!
            )
        }

        // 3- Fused Location Provider
        fusedLocationManager.requestLocationUpdates(
            buildFusedLocationRequest(),
            fusedLocationCallBack as LocationCallback,
            Looper.getMainLooper()
        )

        /*
         Finally : close updated after coroutine scope closed
         */
        awaitClose {
            stopLocationUpdates()
        }
    }

    @SuppressLint("MissingPermission")
    fun getCurrentLocation(): Task<Location?> {
        return fusedLocationManager.getCurrentLocation(
            Priority.PRIORITY_HIGH_ACCURACY,
            CancellationTokenSource().token
        )
    }

    fun stopLocationUpdates() {
        networkLocationListener?.let {
            locationManager.removeUpdates(it)
            networkLocationListener = null
        }

        gpsLocationListener?.let {
            locationManager.removeUpdates(it)
            gpsLocationListener = null
        }
        fusedLocationCallBack?.let {
            fusedLocationManager.removeLocationUpdates(it)
            fusedLocationCallBack = null
        }
    }

    // Build a location request
    private fun buildFusedLocationRequest(): LocationRequest {
        return LocationRequest.create().apply {
            // Sets the desired interval for active location updates.
            // This interval is inexact.
            interval = TimeUnit.SECONDS.toMillis(UPDATE_INTERVAL_SEC)
            // Sets the fastest rate for active location updates.
            // This interval is exact, and the application will never
            // receive updates more frequently than this value
            fastestInterval = TimeUnit.SECONDS.toMillis(FASTEST_UPDATE_INTERVAL_SEC)
            // Sets the maximum time when batched location updates are delivered.
            // Updates may be delivered sooner than this interval
            maxWaitTime = TimeUnit.SECONDS.toMillis(MAXIMUM_WAIT_TIME_SEC)
            // Sets the minimum displacement between location updates in meters
            smallestDisplacement = SMALLEST_DISPLACEMENT_METER
            // Sets the priority of the request.
            priority = Priority.PRIORITY_HIGH_ACCURACY
        }
    }

    companion object {
        private const val UPDATE_INTERVAL_SEC = 30L
        private const val FASTEST_UPDATE_INTERVAL_SEC = 10L
        private const val MAXIMUM_WAIT_TIME_SEC = 60L
        private const val SMALLEST_DISPLACEMENT_METER = 10F
    }
}